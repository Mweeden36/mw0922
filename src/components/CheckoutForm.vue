<template>
  <div class="checkout-form__wrapper panel">
    <div class="checkout-form__item">
      <b class="checkout-form__label">Select a tool: </b>
      <VueMultiselect
        class="tool-selector"
        :allowEmpty="false"
        deselectLabel=""
        label="display"
        :options="toolOptions"
        placeholder="Select a tool to checkout"
        trackBy="code"
        v-model="selectedOption"
      >
        <template v-slot:option="props">
          <ToolCard
            :brand="props.option.brand"
            :code="props.option.code"
            :toolType="props.option.type"
          />
        </template>
      </VueMultiselect>
    </div>
    <div class="checkout-form__item">
      <RentalDates @input="onChangeDates" />
    </div>
    <div class="checkout-form__item">
      <DiscountCounter v-model="discountPercent" />
    </div>
    <div class="checkout-form__item">
      <button class="button checkout-button" @click="onClickCheckout">
        Checkout
      </button>
    </div>
  </div>
</template>

<script>
import "@vuepic/vue-datepicker/dist/main.css";
import VueMultiselect from "vue-multiselect";
import DiscountCounter from "@/components/DiscountCounter.vue";
import RentalDates from "@/components/RentalDates.vue";
import ToolCard from "@/components/ToolCard.vue";
import tools from "@/data/tools.json";
import rentalChargesMixin from "@/mixins/rentalCharges";

export default {
  name: "CheckoutForm",

  components: {
    DiscountCounter,
    RentalDates,
    ToolCard,
    VueMultiselect,
  },

  mixins: [rentalChargesMixin],

  data() {
    return {
      selectedOption: null,
      daysToRent: 0,
      discountPercent: 0,
      rentalStart: null,
      rentalEnd: null,
    };
  },

  computed: {
    discountAmount() {
      return Number((this.subtotal * (this.discountPercent / 100)).toFixed(2));
    },

    discountIsValid() {
      return 100 >= this.discountPercent && this.discountPercent >= 0;
    },

    rentalAgreementInfo() {
      return {
        toolCode: this.selected.code,
        checkoutDate: this.formatDate(this.rentalStart),
        returnDate: this.formatDate(this.rentalEnd),
        discountPercent: this.discountPercent,
        chargeableDays: this.daysToRent,
        dailyCharge: this.dailyRate.toFixed(2),
        prediscountAmount: this.subtotal.toFixed(2),
        discountAmount: this.discountAmount.toFixed(2),
        finalAmount: this.total.toFixed(2),
      };
    },

    selected() {
      /* There's a bug with vue-multiselect where the placeholder text
       *  won't show up if the value is {}. Since we want to be able to
       *  access selected.code in a computed prop, I do this to prevent
       *  an error */
      return this.selectedOption ? this.selectedOption : {};
    },

    subtotal() {
      return this.dailyRate * this.daysToRent;
    },

    toolOptions() {
      return tools.map((tool) => {
        return {
          ...tool,
          display: `${tool.brand} ${tool.type}`,
        };
      });
    },

    toolType() {
      return this.selected.type;
    },

    total() {
      return this.subtotal - this.discountAmount;
    },

    valid() {
      return (
        this.daysToRent > 0 &&
        this.selectedOption !== null &&
        this.discountIsValid
      );
    },
  },

  methods: {
    formatDate(dateVal) {
      const fullYear = dateVal.getFullYear();
      const shortYear = fullYear.toString().substring(2);
      const month = this.prependZero(dateVal.getMonth() + 1);
      const day = this.prependZero(dateVal.getDate());
      return `${month}/${day}/${shortYear}`;
    },

    onChangeDates(data) {
      // Loop over the rental date keys and sync them up.
      ["rentalStart", "rentalEnd", "daysToRent"].forEach((key) => {
        this[key] = data[key];
      });
    },

    onClickCheckout() {
      if (this.valid) {
        // This is typically where I would send the agreement info to the server.
        // If the response is successful, I'd emit this event. Otherwise I'd have some
        // error messaging for the user.
        console.log(this.rentalAgreementInfo);
        return this.$emit("confirm", this.rentalAgreementInfo);
      }
      alert("Please double check your selections");
    },

    prependZero(val) {
      if (val < 10) {
        return `0${val}`;
      }

      return val.toString();
    },
  },
};
</script>

<style src="vue-multiselect/dist/vue-multiselect.css"></style>
<style lang="scss" scoped>
@import "../styles/borders";

.checkout-form__wrapper {
  width: 500px;
  .checkout-form__item {
    padding: 15px 0px;
    .tool-selector {
      display: inline-block;
    }
    :deep(.checkout-form__label) {
      display: block;
      padding-bottom: 10px;
    }
  }
}
</style>
