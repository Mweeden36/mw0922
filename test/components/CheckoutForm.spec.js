import CheckoutForm from "../../src/components/CheckoutForm.vue";
import { describe, it, expect } from "vitest";
import { shallowMount } from "@vue/test-utils";

describe("CheckoutForm.vue", () => {
  const start = new Date();
  const end = new Date(new Date().setDate(start.getDate() + 4));

  it("should initialize with sensible default values", () => {
    const { vm } = shallowMount(CheckoutForm);
    expect(vm.selected).toEqual({});
    expect(vm.daysToRent).toEqual(0);
    expect(vm.discountPercent).toEqual(0);
    expect(vm.rentalStart).toEqual(null);
    expect(vm.rentalEnd).toEqual(null);
    expect(vm.discountAmount).toEqual(0);
    expect(vm.discountIsValid).toEqual(true);
    expect(vm.subtotal).toEqual(0);
    expect(vm.toolOptions).toEqual([
      {
        brand: "Stihl",
        code: "CHNS",
        display: "Stihl chainsaw",
        type: "chainsaw",
      },
      {
        brand: "Werner",
        code: "LADW",
        display: "Werner ladder",
        type: "ladder",
      },
      {
        type: "jackhammer",
        code: "JAKD",
        brand: "DeWalt",
        display: "DeWalt jackhammer",
      },
      {
        type: "jackhammer",
        code: "JAKR",
        brand: "Ridgid",
        display: "Ridgid jackhammer",
      },
    ]);
    expect(vm.toolType).toEqual(undefined);
    expect(vm.total).toEqual(0);
    expect(vm.valid).toEqual(false);
    expect(vm.rentalAgreement).toEqual();
  });

  it("should generate the correct html", () => {
    const wrapper = shallowMount(CheckoutForm);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should prepend zeros to values < 10", () => {
    const { vm } = shallowMount(CheckoutForm);
    for (let i = 0; i <= 9; i++) {
      expect(vm.prependZero(i)).toEqual(`0${i}`);
    }
  });

  it("should handle date updates", () => {
    const { vm } = shallowMount(CheckoutForm);
    expect(vm.daysToRent).toEqual(0);
    expect(vm.rentalStart).toEqual(null);
    expect(vm.rentalEnd).toEqual(null);
    vm.onChangeDates({
      daysToRent: 4,
      rentalStart: start,
      rentalEnd: end,
    });
    expect(vm.daysToRent).toEqual(4);
    expect(vm.rentalStart).toEqual(start);
    expect(vm.rentalEnd).toEqual(end);
  });

  it("should format dates properly", () => {
    const { vm } = shallowMount(CheckoutForm);
    const d = new Date(2019, 5, 3);
    expect(vm.formatDate(d)).toEqual("06/03/19");
  });

  it("should emit an event when checking out", () => {
    const wrapper = shallowMount(CheckoutForm);
    wrapper.vm.selectedOption = {
      code: "JAKD",
      type: "jackhammer",
    };
    wrapper.vm.rentalStart = start;
    wrapper.vm.rentalEnd = end;
    wrapper.vm.daysToRent = 4;

    wrapper.vm.onClickCheckout();
    expect(wrapper.emitted()).toHaveProperty("confirm");
    expect(wrapper.emitted("confirm")[0]).toEqual([
      {
        chargeableDays: 4,
        checkoutDate: expect.any(String),
        dailyCharge: "2.99",
        discountAmount: "0.00",
        discountPercent: 0,
        finalAmount: "11.96",
        prediscountAmount: "11.96",
        returnDate: expect.any(String),
        toolCode: "JAKD",
      },
    ]);
  });
});
