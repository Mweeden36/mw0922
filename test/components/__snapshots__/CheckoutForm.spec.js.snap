// Vitest Snapshot v1

exports[`CheckoutForm.vue > should generate the correct html 1`] = `
"<div class=\\"checkout-form__wrapper panel\\" data-v-5f25d85e=\\"\\">
  <div class=\\"checkout-form__item\\" data-v-5f25d85e=\\"\\"><b class=\\"checkout-form__label\\" data-v-5f25d85e=\\"\\">Select a tool: </b>
    <vue-multiselect-stub name=\\"\\" selectlabel=\\"Press enter to select\\" selectgrouplabel=\\"Press enter to select group\\" selectedlabel=\\"Selected\\" deselectlabel=\\"\\" deselectgrouplabel=\\"Press enter to deselect group\\" showlabels=\\"true\\" limit=\\"99999\\" maxheight=\\"300\\" limittext=\\"(count) => \`and \${count} more\`\\" loading=\\"false\\" disabled=\\"false\\" opendirection=\\"\\" shownooptions=\\"true\\" shownoresults=\\"true\\" tabindex=\\"0\\" class=\\"tool-selector\\" allowempty=\\"false\\" label=\\"display\\" options=\\"[object Object],[object Object],[object Object],[object Object]\\" placeholder=\\"Select a tool to checkout\\" trackby=\\"code\\" data-v-5f25d85e=\\"\\"></vue-multiselect-stub>
  </div>
  <div class=\\"checkout-form__item\\" data-v-5f25d85e=\\"\\">
    <rental-dates-stub data-v-5f25d85e=\\"\\"></rental-dates-stub>
  </div>
  <div class=\\"checkout-form__item\\" data-v-5f25d85e=\\"\\">
    <discount-counter-stub modelvalue=\\"0\\" data-v-5f25d85e=\\"\\"></discount-counter-stub>
  </div>
  <div class=\\"checkout-form__item\\" data-v-5f25d85e=\\"\\"><button class=\\"button checkout-button\\" data-v-5f25d85e=\\"\\"> Checkout </button></div>
</div>"
`;
